bloptools/bayesian/__init__.py

```diff
@@ -288,13 +247,153 @@ def tell(self, new_table=None, append=True, train=True, **kwargs):
+            )
+
+            x = candidates.numpy().astype(float)
+
+            active_X = x[..., [dof["kind"] == "active" for dof in self._subset_dofs(mode="on")]]
+            passive_X = x[..., [dof["kind"] != "active" for dof in self._subset_dofs(mode="on")]]
+            acq_func_meta["passive_values"] = passive_X
+
+        else:
+            if acq_func_name == "quasi-random":
+                active_X = self._subset_inputs_sampler(n=n, kind="active", mode="on").squeeze(1).numpy()
+                acq_func_meta = {"name": "quasi-random", "args": {}}
+
+            else:
+                raise ValueError()
+
+        acq_func_meta["duration"] = duration = ttime.monotonic() - start_time
+
+        if self.verbose:
+            print(f"found points {active_X} in {duration:.01f} seconds")
+
+        if route and n > 1:
+            active_X = active_X[utils.route(self._read_subset_devices(kind="active", mode="on"), active_X)]
+
+        return active_X, acq_func_meta
+
+    def acquire(self, active_inputs):
+        """
+        Acquire and digest according to the self's acquisition and digestion plans.
+
+        This should yield a table of sampled tasks with the same length as the sampled inputs.
+        """
+        try:
+            active_devices = self._subset_devices(kind="active", mode="on")
+            passive_devices = [*self._subset_devices(kind="passive"), *self._subset_devices(kind="active", mode="off")]
+
+            uid = yield from self.acquisition_plan(
+                active_devices,
+                active_inputs.astype(float),
+                [*self.dets, *passive_devices],
+                delay=self.trigger_delay,
+            )
+
+            products = self.digestion(self.db, uid)
+
+            # compute the fitness for each task
+            # for index, entry in products.iterrows():
+            #     for task in self.tasks:
+            #         products.loc[index, task["key"]] = getattr(entry, task["key"])
+
+        except Exception as error:
+            if not self.allow_acquisition_errors:
+                raise error
+            logging.warning(f"Error in acquisition/digestion: {repr(error)}")
+            products = pd.DataFrame(active_inputs, columns=self._subset_dof_names(kind="active", mode="on"))
+            for task in self.tasks:
+                products.loc[:, task["key"]] = np.nan
+
+        if not len(active_inputs) == len(products):
+            raise ValueError("The table returned by the digestion must be the same length as the sampled inputs!")
+
+        return products
+
+    def learn(
+        self,
+        acq_func,
+        n=1,
+        iterations=1,
+        upsample=1,
+        train=True,
+        data=None,
+        **kwargs,
+    ):
+        """
+        This iterates the learning algorithm, looping over ask -> acquire -> tell.
+        It should be passed to a Bluesky RunEngine.
+        """
+
+        if data is not None:
+            if type(data) == str:
```

Comment:
```suggestion
            if type(data) is str:
```

Reply:
Should we also check that the data is a valid file path?

Reply:
I think `pd.read_hdf` on the next line should handle the error ok?